govcd/user.go

```diff
@@ -0,0 +1,456 @@
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		AdminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
+	if willRefresh {
+		err := adminOrg.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, orgUser := range adminOrg.AdminOrg.Users.User {
+		if orgUser.Name == identifier || orgUser.ID == identifier {
+
+			user := NewUser(adminOrg.client, adminOrg)
+
+			_, err := adminOrg.client.ExecuteRequest(orgUser.HREF, http.MethodGet,
+				types.MimeAdminUser, "error getting user: %s", nil, user.User)
+
+			return user, err
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetRole finds a role within the organization
+func (adminOrg *AdminOrg) GetRole(roleName string) (*types.Reference, error) {
+
+	// There is no need to refresh the AdminOrg, until we implement CRUD for roles
+	for _, role := range adminOrg.AdminOrg.RoleReferences.RoleReference {
+		if role.Name == roleName {
+			return role, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// CreateUser creates an OrgUser from a full configuration structure
+// The timeOut variable is the maximum time we wait for the user to be ready
+// (This operation does not return a task)
+// This function returns as soon as the user has been created, which could be as
+// little as 200ms or as much as Client.MaxRetryTimeout
+// Mandatory fields are: Name, Role, Password.
+// https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-CreateUser.html
+func (adminOrg *AdminOrg) CreateUser(userConfiguration *types.User) (*OrgUser, error) {
```

Comment:
I think we need to check that user is sysAdmin

Reply:
Isn't this implicit when we use an AdminOrg?

Reply:
I guess in this case you are right.